### PR TITLE
Add Support for Qwen3-30B-A3B and Qwen3-Coder-480B-A35B MoE Models

### PR DESCRIPTION
**Pull request by parambole**
#### **TL;DR**

-   **What**: This PR adds support for two new Qwen3 Mixture-of-Experts (MoE) models: **`Qwen3-30B-A3B`** and **`Qwen3-Coder-480B-A35B`**.

-   **Why**: To expand the range of supported MoE models within the MaxText ecosystem.

-   **How**: By adding their respective model configuration YAML files, updating the checkpoint conversion utility with their architectural parameters, and updating the project documentation.

* * * * *

#### **Detailed Description**

This PR integrates two additional variants of the Qwen3 MoE model family into MaxText, enabling training and inference for these architectures.

#### **Key Changes**

1.  **Model Configurations (`configs/models/`)**:

    -   Added `qwen3-30b-a3b.yml` to define the architecture for the 30B parameter model.

    -   Added `qwen3-480b-a35b.yml` to define the architecture for the 480B parameter coder model.

2.  **Checkpoint Converter Update (`convert_qwen3_moe.py`)**:

    -   The `HF_QWEN3_MOE_CONFIG` dictionary has been updated with the specific architectural parameters for both `qwen3-30b-a3b` and `qwen3-480b-a35b`, ensuring the checkpoint conversion script can correctly handle these models.

3.  **Framework Integration (`pyconfig.py`)**:

    -   The new model names have been added to the `validate_model_name` function to allow them to be recognized by the MaxText framework.

4.  **Documentation (`README.md`)**:

    -   The project `README.md` has been updated to announce the official support for the newly added Qwen3 MoE models.

* * * * *

#### **Testing Strategy & Validation**

The correctness of the implementation was validated by performing a **forward logit verification** for both models. The MaxText implementation's output logits were compared against the original Hugging Face models, confirming that the probability distributions are correctly aligned.

##### ✅ **Validation for Qwen3-30B-A3B**

```
--- Prompt: What is the ---

--- MaxText model top 10 tokens ---
| Token ID   | Token                | Score      |
|------------|----------------------|------------|
| 829        | name                 | 25.0000    |
| 2265       | title                | 24.6250    |
| 1887       | main                 | 24.2500    |
| 25361      | significance         | 23.8750    |
| 3476       | role                 | 23.8750    |
| 7428       | purpose              | 23.7500    |
| 5025       | relationship         | 23.3750    |
| 6672       | difference           | 23.2500    |
| 6028       | primary              | 23.0000    |
| 14806      | formula              | 22.6250    |

--- HF model top 10 tokens ---
| Token ID   | Token                | Score      |
|------------|----------------------|------------|
| 829        | name                 | 25.0000    |
| 2265       | title                | 24.6250    |
| 1887       | main                 | 24.2500    |
| 3476       | role                 | 23.8750    |
| 25361      | significance         | 23.8750    |
| 7428       | purpose              | 23.7500    |
| 5025       | relationship         | 23.3750    |
| 6672       | difference           | 23.2500    |
| 6028       | primary              | 23.1250    |
| 14806      | formula              | 22.6250    |

--- Similarity Metrics of Top Tokens ---
| Metric                       | Value                |
|------------------------------|----------------------|
| overlap_count                | 10/10                |
| jaccard_similarity           | 1.0                  |
| rank_agreement_percentage    | 80.0                 |

Average KL divergence per token (D_KL(P_golden || Q_model)): 0.001571
Max KL divergence for a single token in the set: 0.003658

```

##### ✅ **Validation for Qwen3-Coder-480B-A35B**

```
--- Prompt: I love to ---

--- MaxText model top 10 tokens ---
| Token ID   | Token                | Score      |
|------------|----------------------|------------|
| 3736       | watch                | 11.1875    |
| 4538       | teach                | 11.1250    |
| 5821       | travel               | 11.1250    |
| 1490       | see                  | 11.0625    |
| 1349       | read                 | 10.7500    |
| 3270       | write                | 10.6875    |
| 15254      | dance                | 10.6875    |
| 1281       | make                 | 10.5000    |
| 4296       | cook                 | 10.0625    |
| 1486       | play                 | 10.0625    |

--- HF model top 10 tokens ---
| Token ID   | Token                | Score      |
|------------|----------------------|------------|
| 5821       | travel               | 11.0625    |
| 3736       | watch                | 11.0000    |
| 4538       | teach                | 10.9375    |
| 1490       | see                  | 10.8750    |
| 1349       | read                 | 10.5625    |
| 15254      | dance                | 10.5000    |
| 3270       | write                | 10.5000    |
| 1281       | make                 | 10.3125    |
| 4296       | cook                 | 9.9375     |
| 8180       | eat                  | 9.8750     |

--- Similarity Metrics of Top Tokens ---
| Metric                       | Value                |
|------------------------------|----------------------|
| overlap_count                | 9/10                 |
| jaccard_similarity           | 0.8181818181818182   |
| rank_agreement_percentage    | 40.0                 |

Average KL divergence per token (D_KL(P_golden || Q_model)): 0.002180
Max KL divergence for a single token in the set: 0.006329

```

* * * * *

#### **Checklist**

-   [X] I have performed a self-review of my code.
-   [X] I have necessary comments in my code, particularly in hard-to-understand areas.
-   [X] I have run end-to-end tests and provided the validation results above.
-   [X] I have made or will make corresponding changes to the documentation if needed.

### CODE REVIEW

File: MaxText/configs/models/qwen3-480b-a35b.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+mlp_activations: ["silu", "linear"]
+vocab_size: 151936
+normalization_layer_epsilon: 1.0e-6
+use_qk_norm: true
+
+# MoE Specific Parameters
+num_experts: 160
+num_experts_per_tok: 8
+base_moe_mlp_dim: 2560
+load_balance_loss_weight: 0.0
```
Comment (gagika) [inline MaxText/configs/models/qwen3-480b-a35b.yml]: why is it 0?

File: MaxText/configs/models/qwen3-30b-a3b.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+mlp_activations: ["silu", "linear"]
+vocab_size: 151936
+normalization_layer_epsilon: 1.0e-6
+use_qk_norm: True
+
+# MoE Specific Parameters
+num_experts: 128
+num_experts_per_tok: 8
+base_moe_mlp_dim: 768
+load_balance_loss_weight: 0.001
```
Comment (gagika) [inline MaxText/configs/models/qwen3-30b-a3b.yml]: shouldn't `load_balance_loss_weight` be part of training recipe and not model config?

File: MaxText/configs/models/qwen3-480b-a35b.yml (hunk 1 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-480b-a35b.yml]: Hey Gagik, the [router_aux_loss_coef](https://huggingface.co/Qwen/Qwen3-Coder-480B-A35B-Instruct/blob/main/config.json#L29) is set to 0.0.

File: MaxText/configs/models/qwen3-30b-a3b.yml (hunk 2 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-30b-a3b.yml]: But you are right. This should not be something that influences the forward pass and, hence, should not be part of the model config. It is better suited in the training recipe. I have removed it from here and other places.

**Review by gagika — APPROVED** [review]
Thanks

File: MaxText/configs/models/qwen3-30b-a3b.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+num_experts: 128
+num_experts_per_tok: 8
+base_moe_mlp_dim: 768
+norm_topk_prob: true
+
+# RoPE Settings
+rope_max_timescale: 10000000
+
+# General Model Settings
+enable_dropout: False
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-30b-a3b.yml]: Nit: extra line.

File: MaxText/configs/models/qwen3-30b-a3b.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+use_qk_norm: True
+
+# MoE Specific Parameters
+num_experts: 128
+num_experts_per_tok: 8
+base_moe_mlp_dim: 768
+norm_topk_prob: true
+
+# RoPE Settings
+rope_max_timescale: 10000000
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-30b-a3b.yml]: Nit: `10_000_000`. Same for other configs.

File: MaxText/configs/models/qwen3-30b-a3b.yml (hunk 3 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-30b-a3b.yml]: Done. Thanks

File: MaxText/configs/models/qwen3-30b-a3b.yml (hunk 4 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-30b-a3b.yml]: Done. Thanks

File: MaxText/configs/models/qwen3-30b-a3b.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+head_dim: 128
+mlp_activations: ["silu", "linear"]
+vocab_size: 151936
+normalization_layer_epsilon: 1.0e-6
+use_qk_norm: True
+
+# MoE Specific Parameters
+num_experts: 128
+num_experts_per_tok: 8
+base_moe_mlp_dim: 768
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-30b-a3b.yml:33]: Please update this to `base_mlp_dim`, and add an assertion in pyconfig.py file like `validate_mlp_dim()` with:

```
if first_num_dense_layers == 0:
    assert base_moe_mlp_dim == base_mlp_dim, (
        "When `first_num_dense_layers` is 0, `base_moe_mlp_dim` must be equal to `base_mlp_dim`."
    )
 ```

**Review by RissyRan — COMMENTED** [review]
Discussed offline. A small change needed

**Review by RissyRan — CHANGES_REQUESTED** [review]
no-op

File: MaxText/configs/models/qwen3-30b-a3b.yml (hunk 5 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-30b-a3b.yml:33]: Done. Thank you for catching this.

**Review by RissyRan — APPROVED** [review]
Thank you for the change!

File: MaxText/configs/models/llama4-17b-128e.yml [hunk 6]
```diff
@@ -22,7 +22,7 @@ base_emb_dim: 5120
 base_num_decoder_layers: 48
 base_num_query_heads: 40
 base_num_kv_heads: 8
-base_mlp_dim: 16384
+base_mlp_dim: 8192
```
Comment (RissyRan) [inline MaxText/configs/models/llama4-17b-128e.yml]: Hmm... Llama4 is tricky, we cannot do this change here, especially `interleave_moe_layer_step: 2` for llama4-17b-128e.

```
"intermediate_size": 8192,
"intermediate_size_mlp": 16384,
```

I'd like to test all moe models with this change (flops before vs. after, take it carefully). I think this ask is out of this PR scope. So we could:

1) In this PR, just update Qwen3 from base_moe_dim to base_mlp_dim;
2) Cut a bug and assign to me to add this assertion for future MoE models (or if you'd like to follow it up, I am happy to review).

**Review by richjames0 — APPROVED** [review]
lgtm

File: MaxText/configs/models/qwen3-30b-a3b.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Model config for Qwen3-30B-A3B
+
+# Core Architectural Parameters
+decoder_block: "qwen3_moe"
+base_emb_dim: 768
+base_moe_mlp_dim: 768
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-30b-a3b.yml]: Hey, let's double check :) 

```
"hidden_size": 2048, --> emb_dim
"moe_intermediate_size": 768,--> base_mlp_dim not base_moe_mlp_dim
```

Comment (parambole) [reply MaxText/configs/models/qwen3-30b-a3b.yml]: Done. Thanks for catching this.

File: MaxText/configs/models/llama4-17b-128e.yml (hunk 6 above)
Comment (parambole) [reply MaxText/configs/models/llama4-17b-128e.yml]: b/440352365
